docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+       {
+           if (requested_formats[n] == attester_formats[m])
+               selected_format = &attester_formats[m];
+       }
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
```

Comment:
No idea what "oe_sgx_get_verifiers" is here, this is mentioned without any previous explanation.  Need more context, since offhand that sounds like an API we should not have (because the name is SGX specific).

Also I suspect "oe_register_verifier" is misnamed and should be "oe_register_verifier_plugin".

Reply:
Added oe_sgx_get_verifiers in motivation. This is a replacement of current SGX-specific API.

oe_register_verifier() is an existing function.


---
In reply to: [393251219](https://github.com/openenclave/openenclave/pull/2621#discussion_r393251219) [](ancestors = 393251219)

Reply:
I have not seen justification for replacing a current SGX-specific API with another SGX-specific API as opposed to replacing it with a TEE-agnostic API.   So I don't see my question as answered yet.

Reply:
Renamed
. oe_sgx_get_attesters to oe_get_attesters
. oe_sgx_get_verifiers to oe_get_verifiers

---
In reply to: [394652466](https://github.com/openenclave/openenclave/pull/2621#discussion_r394652466) [](ancestors = 394652466)